### PR TITLE
fix: Allow multiple path segments in BasePath

### PR DESCRIPTION
**Pull request by agrippa1994**
*Issue #, if available:*
None

*Description of changes:*
The API gateway supports multiple segments in the base path mappings like specified in this [repository](https://github.com/aws-samples/sessions-with-aws-sam/blob/master/multi-level-mapping/reportingv1.yaml#L21). However, whenever a ```BasePath``` with multiple segments is specified in a ```AWS::Serverless::HttpApi``` (see examples below), than the template does not contain these multiple segments after the transformation anymore.

```yaml
# SAM Template:
HttpApi:
    Type: AWS::Serverless::HttpApi
    Properties:
      Domain:
        DomainName: somedomain.com
        CertificateArn: some-certificate-arn
        BasePath:
          - "api/v1"
        EndpointConfiguration: REGIONAL
        Route53:
          HostedZoneId: some-hosted-zone
```
```json
// transformed template
"HttpApiapiApiMapping": {
  "Type": "AWS::ApiGatewayV2::ApiMapping",
  "Properties": {
    "ApiId": {
      "Ref": "HttpApi"
    },
    "ApiMappingKey": "api",
    "DomainName": {
      "Ref": "ApiGatewayDomainNameV2dc0129008b"
    },
    "Stage": {
      "Ref": "HttpApiApiGatewayDefaultStage"
    }
  }
}
```
The implementation fixes the regex that strips trailing and leading forward slashes (the segments). After that, the transformed template looks as following example:
```json
// transformed template
"HttpApiapiv1ApiMapping": {
  "Type": "AWS::ApiGatewayV2::ApiMapping",
  "Properties": {
    "ApiId": {
      "Ref": "HttpApi"
    },
    "ApiMappingKey": "api/v1",
    "DomainName": {
      "Ref": "ApiGatewayDomainNameV2dc0129008b"
    },
    "Stage": {
      "Ref": "HttpApiApiGatewayDefaultStage"
    }
  }
},
```

*Description of how you validated changes:*
I used my SAM template with a ```BasePath``` with multiple segments. After transformation and deployment, the ApiMappingKey was correct for my Domain.

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Update documentation
- [x] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by hoffa** [comment]
Thanks for the PR! Likely related: https://github.com/aws/serverless-application-model/pull/2041

**Review by hoffa — APPROVED** [review]

**Review by c2tarun — APPROVED** [review]

File: samtranslator/model/api/http_api_generator.py [hunk 1]
```diff
@@ -344,12 +344,9 @@ def _construct_basepath_mappings(self, basepaths, http_api):
                     path = ""
                 else:
                     # ignore leading and trailing `/` in the path name
-                    m = re.search(r"[a-zA-Z0-9]+[\-\_]?[a-zA-Z0-9]+", path)
-                    path = m.string[m.start(0) : m.end(0)]
-                    if path is None:
-                        raise InvalidResourceException(self.logical_id, "Invalid Basepath name provided.")
+                    path = re.sub(r"^/+|/+$", "", path)
```
Comment (mgrandis) [inline samtranslator/model/api/http_api_generator.py:347]: I think we could replace the whole `if / else` by a `path.strip("/")`

File: tests/model/api/test_http_api_generator.py [hunk 2]
```diff
@@ -250,15 +250,18 @@ def test_basepaths(self):
         self.kwargs["domain"] = {
             "DomainName": "example.com",
             "CertificateArn": "some-url",
-            "BasePath": ["one-1", "two_2", "three", "/"],
+            "BasePath": ["one-1", "two_2", "three", "/", "api/v1"],
```
Comment (mgrandis) [inline tests/model/api/test_http_api_generator.py:253]: We should add: `"/api"`, `"api/v1/"`, `"/api/v1/"` cases to test the `/` stripping.

**Review by mgrandis — CHANGES_REQUESTED** [review]
Just a few remarks

**Comment by agrippa1994** [comment]
Thanks for your review. I'm going to fix them as soon as I have some time for doing that.

**Comment by agrippa1994** [comment]
@mgrandis I have implemented your requested changes. The stripping method works perfectly fine here.

File: samtranslator/model/api/http_api_generator.py [hunk 3]
```diff
@@ -344,12 +344,9 @@ def _construct_basepath_mappings(self, basepaths, http_api):
                     path = ""
                 else:
                     # ignore leading and trailing `/` in the path name
-                    m = re.search(r"[a-zA-Z0-9]+[\-\_]?[a-zA-Z0-9]+", path)
-                    path = m.string[m.start(0) : m.end(0)]
-                    if path is None:
-                        raise InvalidResourceException(self.logical_id, "Invalid Basepath name provided.")
+                    path = path.strip("/")
```
Comment (mgrandis) [inline samtranslator/model/api/http_api_generator.py:347]: I think you can actually replace the whole if/else.
if `path == "/"`, `path.strip("/")` will result in `""`.
So you can replace the whole L343-L347 with `path = path.strip("/")`

**Review by mgrandis — APPROVED** [review]

**Pull request merged**